File: mqtt_chat_node/chat_echo_server.js
#!/usr/bin/env node
//
// MQTT Chat loopback test server
//
// Author : Maho Takara  takara@jp.ibm.com
// 2016/11/26  Initial commit
//
var mqtt    = require('mqtt');
var cnf = require('./config.json');
var client  = mqtt.connect(cnf.broker_url);


// Callback as the connection established
client.on('connect', function () {
    console.log("connect & subscribe");
    client.subscribe(cnf.broker_topic);
});

// Callbask as the message received
client.on('message', function (topic, message) {
    console.log("=== receive message ===");
    console.log("topic: ", topic.toString());
    console.log("message: ", message.toString());
    // convert from text to json
    json_message = JSON.parse(message);
    console.log(json_message);
    client.publish(json_message.reply, message.toString());
});

File: mqtt_chat_ruby/chat_echo_server.rb
#!/usr/bin/env ruby 
# encoding: utf-8
#
# MQTT Echo back server
# 
# Author : Maho Takara  takara@jp.ibm.com
# 2016/11/26  Initial commit

require 'mqtt'
require 'json'


def mqtt_loop(client,cnf)
  client.get(cnf['listen_topic']) do |topic,message|
    puts "#{topic},#{message}"
    
    hash_message = JSON.parse(message)
    print "hash reply = ", hash_message['reply'],"\n"
    client.publish(bash_message['reply'], message, retain=false)
  end
end


##### MAIN #####
if __FILE__ == $0

  cnf = open("config.json") do |io|
    JSON.load(io)
  end

  client = MQTT::Client.connect(cnf['mqtt_broker'])
  print "waiting message\n"
  mqtt_loop(client,cnf)
  client.disconnect()

end
